src/Sarif.Viewer.VisualStudio/Controls/InfoBar.cs

```diff
@@ -45,32 +47,52 @@ public class InfoBar : IVsInfoBarUIEvents
         private IVsInfoBarUIElement uiElement;
         private uint eventCookie;
 
+        private static readonly ReaderWriterLockSlimWrapper s_infoBarLock = new ReaderWriterLockSlimWrapper(new ReaderWriterLockSlim());
+
+        // It might seem more natural to use the condition as the dictionary key, and the InfoBar
+        // as the value. We wrote it this way because when the user closes an InfoBar, it's easier
+        // to remove it from the dictionary if the InfoBar itself is the key -- rather than having
+        // to look up the KeyValuePair that has that InfoBar as its value. See CloseAsync below.
+        private static readonly IDictionary<InfoBar, ExceptionalConditions> s_infoBarToConditionDictionary = new ConcurrentDictionary<InfoBar, ExceptionalConditions>();
+
         /// <summary>
         /// Display info bars appropriate to the specified set of "exceptional conditions."
         /// </summary>
         /// <param name="conditions">
         /// The conditions that require an info bar to be shown.
         /// </param>
-        internal static void CreateInfoBarsForExceptionalConditions(ExceptionalConditions conditions)
+        internal static async Task CreateInfoBarsForExceptionalConditionsAsync(ExceptionalConditions conditions)
         {
-            if ((conditions & ExceptionalConditions.InvalidJson) != 0)
-            {
-                new InfoBar(Resources.ErrorInvalidSarifStream).ShowAsync().FileAndForget(FileAndForgetEventName.InfoBarOpenFailure);
-            }
-
-            if ((conditions & ExceptionalConditions.ConfigurationError) != 0)
-            {
-                new InfoBar(Resources.ErrorLogHasErrorLevelToolConfigurationNotifications).ShowAsync().FileAndForget(FileAndForgetEventName.InfoBarOpenFailure);
-            }
-
-            if ((conditions & ExceptionalConditions.ExecutionError) != 0)
-            {
-                new InfoBar(Resources.ErrorLogHasErrorLevelToolExecutionNotifications).ShowAsync().FileAndForget(FileAndForgetEventName.InfoBarOpenFailure);
-            }
-
-            if ((conditions & ExceptionalConditions.NoResults) != 0)
+            using (s_infoBarLock.EnterWriteLock())
```

Comment:
>s_infoBarLock.EnterWriteLock() [](start = 19, length = 30)

So this is really holding a lock while calling into "ShowAsync" below.. Any concerns at all about that? (Like ShowAsync taking a while, or even worse, ending up back here?)

Reply:
Yes, and I'll change that. See my chat with Xiang Yu -- the sequence of four calls to `AddInfoBarIfRequiredAsync` doesn't have to be atomic. I only need to hold the lock _within_ each call, to atomicize the logic that determines that I need to access the dictionary and the dictionary access itself. I'll move the `ShowAsync` call outside the lock. Good catch, thanks for upping my `async` game.

---
In reply to: [527329760](https://github.com/microsoft/sarif-visualstudio-extension/pull/278#discussion_r527329760) [](ancestors = 527329760)